Fix zend_alloc aligned allocation on Windows

mvorisek opened the pull request:
originally, on Windows, the memory was aligned by an assumption the original memory is available for immediately following allocation

even if the aligned allocation was inside the original address space, it seems this assumption is not always correct and the aligned allocation might fail, especially in 32-builds where is address space is much smaller

fixes #9650 and https://bugs.php.net/bug.php?id=77194

also no need to dump `errno` on linux twice, it is dumped already when reporting lower level alloc failures (mmap, nmmap, munmap)

cmb69 commented:
Please target the "master" branch.

mvorisek commented on Zend/zend_alloc.c:
```diff
@@ -415,11 +415,38 @@ stderr_last_error(char *msg)
 {
 #ifdef _WIN32
-	return VirtualAlloc(addr, size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+	void *ptr = VirtualAlloc(addr, size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
```
I belive here is the problem causing https://github.com/php/php-src/issues/9650

citing the docs: https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc#remarks

> If the lpAddress parameter is not NULL, the function uses the lpAddress and dwSize parameters to compute the region of pages to be allocated. The current state of the entire range of pages must be compatible with the type of allocation specified by the flAllocationType parameter. Otherwise, the function fails and none of the pages are allocated. This compatibility requirement does not preclude committing an already committed page, as mentioned previously.

On linux, when the add+new_size is occupied, a different pointer (unless MAP_FIXED) is returned. On Windows, it is a failure and we must handle it.

TysonAndre replied:
> zend_mm_chunk_alloc_int

That seems likely - I missed earlier that when zend_mm_init called zend_mm_chunk_alloc_int, zend_mm_chunk_alloc_int would go to unmap then map, just as you mentioned

```c
static void *zend_mm_chunk_alloc_int(size_t size, size_t alignment)
{
	void *ptr = zend_mm_mmap(size);

	if (ptr == NULL) {
		return NULL;
	} else if (ZEND_MM_ALIGNED_OFFSET(ptr, alignment) == 0) {
		if (zend_mm_use_huge_pages) {
			zend_mm_hugepage(ptr, size);
		}
		return ptr;
	} else {
		size_t offset;

		/* chunk has to be aligned */
		zend_mm_munmap(ptr, size);
		ptr = zend_mm_mmap(size + alignment - REAL_PAGE_SIZE);
#ifdef _WIN32
		offset = ZEND_MM_ALIGNED_OFFSET(ptr, alignment);
		zend_mm_munmap(ptr, size + alignment - REAL_PAGE_SIZE);
		ptr = zend_mm_mmap_fixed((void*)((char*)ptr + (alignment - offset)), size);
		offset = ZEND_MM_ALIGNED_OFFSET(ptr, alignment);
		if (offset != 0) {
			zend_mm_munmap(ptr, size);
			return NULL;
		}
		return ptr;
#else
```

TysonAndre commented:
So if VirtualAlloc fails on win32 when realigning the pointer, was your plan to allocate a larger block of size `size + alignment` that wasn't fixed, then align it (e.g. by unmapping extra memory from the start and end where needed)? (That's what I would try)

```
#define ZEND_MM_CHUNK_SIZE ((size_t) (2 * 1024 * 1024))    /* 2 MB  */
```

> https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170
> `_WIN32` Defined as 1 when the compilation target is 32-bit ARM, 64-bit ARM, x86, or x64. Otherwise, undefined.

This is a bug fix for a random crash due to getting unlucky allocating a memory page that affects 32-bit windows, so my preference would be targeting 8.0 myself.

Maybe
 
- Only change the handling of the fixed mapping returning null when `#ifdef _WIN32`
- Change the logging in php's `master` branch for 8.3.0-dev

TysonAndre commented:
https://stackoverflow.com/a/14816824

Are we sure that the page size is actually 4KB at runtime in appveyor? Maybe VirtualAlloc is somehow using a large page size for VirtualAlloc, and as a result it's rounding up to the next 8KB, and as a result it's more likely to request mmap_fixed after the end of a region

https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getsysteminfo can be called to check the page size at runtime

> [in] dwSize
>
>The size of the region, in bytes. If the lpAddress parameter is NULL, this value is rounded up to the next page boundary. Otherwise, the allocated pages include all pages containing one or more bytes in the range from lpAddress to lpAddress+dwSize. This means that a 2-byte range straddling a page boundary causes both pages to be included in the allocated region.

```
#define ZEND_MM_PAGE_SIZE  (4 * 1024)                      /* 4 KB  */
```

TysonAndre commented:
More brainstorming:

https://github.com/microsoft/mimalloc/blob/master/src/os.c#L411-L422 https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc2 - it seems like VirtualAlloc2 supports huge pages as well as passing the requested alignment to the VirtualAlloc2 call itself, but only works in windows 10+ (that seems too new, but it's useful to know about)

Zend/zend.c already defines a helper to get the page size - `ZEND_API size_t zend_get_page_size()` - we could print out the value on appveyor for the x86 platform (`_WIN32`) (e.g. maybe add it to phpinfo or zend_test so we can assert that 32-bit builds have the expected page size of 4kb on windows and not something larger)?

mvorisek commented:
I did some more debugging on x86 Win build:

```
# random test failure I:
XXXX_MM_LOG_XXXX:
alloc: null, 0CCB0000, 200000
free: 0CCB0000, 200000
alloc: null, 0CCB0000, 3ff000
free: 0CCB0000, 3ff000
alloc: 0CE00000, 00000000, 200000

XXXX_MM_LOG_XXXX:
free: 0CE00000, 200000
    VirtualFree() failed: [0x000001e7] Attempt to access invalid address
    Can't initialize heap


# random test failure II:
XXXX_MM_LOG_XXXX:
alloc: null, 0C300000, 200000
free: 0C300000, 200000
alloc: null, 0C300000, 3ff000
free: 0C300000, 3ff000
alloc: 0C400000, 00000000, 200000

XXXX_MM_LOG_XXXX:
free: 0C400000, 200000
    VirtualFree() failed: [0x000001e7] Attempt to access invalid address
    Can't initialize heap

# ext\gd\tests\bug77269.phpt (test needs to be fixed fo x86, unrelated):
alloc: null, 0BF00000, 200000
free: 0BF00000, 200000
alloc: null, 0BF00000, 3ff000
free: 0BF00000, 3ff000
alloc: 0C000000, 0C000000, 200000
alloc: null, 29000000, 10000000
alloc: null, 00000000, 40000000
    VirtualAlloc(NULL) failed: [0x00000008] Not enough memory resources are available to process this command
```

syntax is:
```
alloc: orig_ptr, new_ptr, new_size === new_ptr = VirtualAlloc(orig_ptr, new_size)
free: ptr, size === VirtualFree(ptr, size)
```

PHP zend_alloc is extending the memory by passing the next address and tries to allocate it. This have at least two problems:

## 1. invalid address can be passed if the previous virtual address was on the very end of the valid space

like passing value greater than 0x7FFFFFFF in x86 Win build

## 2. if allocation fails, php currently tries to free the memory (ptr passed to the alloc func), this is wrong, as passed pointer might be not previously allocated (as shown in the log, in the code, we must pass an extra arg to `zend_mm_mmap_fixed` if ptr is wanted or already allocated)

TysonAndre commented:
My best idea for a workaround is to retry the entire request for aligned memory (alloc with extra + free + alloc aligned at pointer) on win32 if it fails, a finite number of times (in case there's actually no memory or another error)

- Probably also including the happy path of requesting ZEND_MM_CHUNK_SIZE=2MB and returning early if it's already aligned to ZEND_MM_CHUNK_SIZE

E.g. 10 times.

If that fails, then return a fatal error like it did before for being unable to allocate memory

-----

For Windows 10, https://github.com/microsoft/mimalloc/blob/master/src/os.c# has an example of calling GetProcAddress to look up the function pointer of VirtualAlloc2, which lets you pass in the alignment hint to the allocator

mvorisek commented:
> immediately freeing it with VirtualFree

it is a "my custom debug log" from zend_alloc, so there are/were some php ops between

> My best idea for a workaround is to retry the entire request for aligned memory

so this is thankfully not needed

> 0x000001e7 Win error

comes from `zend_mm_mmap_fixed` with given != null ptr, but ptr - ptr+size is not avaialble, easy to reproduce on many test runs /w x86 builds

linux has same problem, free may be called on data php did not allocate

I am understanding this problem, so I will soon, then we can discuss if there is anything non-bug related to improve

cmb69 commented:
> I did some more debugging on x86 Win build:

Thank you! (Although I don't think that 32bit Windows builds are important; does anybody still use them?)

> alloc: 0CE00000, 00000000, 200000

That looks fishy. Why would we want to allocate zero bytes in the first place? And I wouldn't be surprised that nothing is allocated in this case, so trying to free should fail.

mvorisek commented:
> Thank you! (Although I don't think that 32bit Windows builds are important; does anybody still use them?)

I do not use 32 bit builds ATM neither, but I belive 32 bit builds are important to discover issues like these, which are present also on 64 bit builds but are much harder to debug because they are much more rare.

> > alloc: 0CE00000, 00000000, 200000

see "syntax is", the middle value is result of VirtualAlloc, all ok, we only need to handle it :)

mvorisek commented:
what looks a little fishy to me is `0x3ff000` size allocation, there is some aligment in the zend_alloc, the original allocation size is `0x200000`, but then I would expect (same nicely rounded) `0x400000` value, is that intended?

TysonAndre commented:
> what looks a little fishy to me is 0x3ff000 size allocation, there is some aligment in the zend_alloc, the original allocation size is 0x200000, but then I would expect (same nicely rounded) 0x400000 value, is that intended?


Yes, that's deliberate - a block of memory of size 2MB + 2MB - 4kb (size + alignment - REAL_PAGE_SIZE is guaranteed to have one region that's of size exactly 2MB if the allocator aligns to the 4kb page size

```
		/* chunk has to be aligned */
		zend_mm_munmap(ptr, size);
		ptr = zend_mm_mmap(size + alignment - REAL_PAGE_SIZE);
```

mvorisek commented:
PR is done, I updated the description to match the real problem

tested with many https://github.com/php/php-src/pull/8392 runs for x86 and x64 Windows builds and it fixes the #9650 reliably, the original asumption about the aligned allocation after larger alloc was simply wrong

TysonAndre commented:
At a glance at the approach - This seems like it'll waste reserved address space on 32 bit builds - it's reserved but not unreserved later, so future calls with the null pointer won't include the wasted reserved space? Does that sound right?

E.g. if a 4MB block is allocated, but it isn't aligned, then 2MB will be **committed**, but 4MB will be **reserved** - on 32-bit Windows bits that'll be a bug because the extra space will stay reserved (and not used in future calls to the allocator) since that'll reduce the fraction of the small 2GB(or 3GB?) of space that can actually be used by applications in 32-bit builds

TysonAndre commented:
https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualfree#parameters

> E.g. if a 4MB block is allocated, but it isn't aligned, then 2MB will be committed, but 4MB will be reserved

VirtualFree doesn't offer a way to unreserve a region, it can only unreserve entire blocks. There's only RELEASE (unreserve+uncommit) and DECOMMIT(make memory stop being committed, but it stays reserved)

-----

As a result of this change, I'd expect that e.g. creating a really large SplDoublyLinkedList (quick test for lots of small allocations with no really large allocations) would fail at a smaller size than it would before this change on win32, running 32-bit php on 64-bit os with more than 4GB of memory

------


The approach of a finite number of retries I'd mentioned earlier seems like the best option.

mvorisek commented:
> VirtualFree doesn't offer a way to unreserve a region, it can only unreserve entire blocks.

yes

> The approach of a finite number of retries I'd mentioned earlier seems like the best option.

several tries of "alloc + dealloc + smaller aligned alloc" fails once the 1st failed, that is almost expected, we will have to store the allocated parts first... and still only a probabilistic "almost never to fail" solution

> This seems like it'll waste reserved address space on 32 bit builds - it's reserved but not unreserved later, so future calls with the null pointer won't include the wasted reserved space? Does that sound right?

I am aware of more used virtual memory space and I will fix this for 32-bit builds to try the allocation as before first (which fails in like 1/10k test run for 32-bit build).

TysonAndre commented:
> several tries of "alloc + dealloc + smaller aligned alloc" fails once the 1st failed, that is almost expected, we will have to store the allocated parts first... and still only a probabilistic "almost never to fail" solution

To confirm my understanding: You're saying you've implemented and tested the implementation with retrying and still got the zend_mm_init failure?


Optional: Do you have logs from that or a branch with that for future reference?



------

At a glance, I assume this introduces a new bug: `VirtualFree(addr, 0)` is called when deallocating, and the page won't be freed - this won't work if the real address is `addr - offset` for passing size 0. If this bug is only encountered at startup, this is less of an issue, but does it also affect long-running php processes?

- https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualfree I'm not certain. I'm assuming the initial call of VirtualAlloc it refers to is the RESERVE, not the COMMIT
- E.g. could try running a ZTS web server with something like `ab` apache bench to make a lot of requests

How does the zend memory manager work with ZTS windows builds? I'm guessing there's a different allocator for each thread - so if this also affects 1 in 10000 HTTP requests to a ZTS web server, then this might slowly leak reserved memory.

- This is better than crashing, I guess?
- I hope it doesn't leak in that way
- It might expose brand new bugs, though - e.g. pecl extensions would see errors more often if they never get allocation errors but call VirtualAlloc and get errors reserving memory
- Slowly leaking reserved memory might cause applications to use high memory addresses more often (I forget if we allow 3GB of allocated memory for 32-bit windows - those pointers would be negative for 0x8000000-0xffffffff , and the allocator normally returns low pointers.)

https://windows.php.net/download/

```c
#ifdef ZTS
static int alloc_globals_id;
static size_t alloc_globals_offset;
# define AG(v) ZEND_TSRMG_FAST(alloc_globals_offset, zend_alloc_globals *, v)
#else
# define AG(v) (alloc_globals.v)
static zend_alloc_globals alloc_globals;
#endif
```

If it's a singleton allocator, this approach is probably fine - I'd assume this bug only affects low addresses and when the process initially starts up.

mvorisek commented:
> At a glance, I assume this introduces a new bug: `VirtualFree(addr, 0)` is called when deallocating, and the page won't be freed - this won't work if the real address is `addr - offset` for passing size 0. If this bug is only encountered at startup, this is less of an issue, but does it also affect long-running php processes?

`VirtualFree` expects base address given for deallocation and size of `0` and the base address is correctly queried

in https://github.com/php/php-src/pull/9721/commits/87a1025404bbc2128f3fd98f0069a3a6efde255c I have added the original approach to be tried on 32 bit build first

TysonAndre commented on Zend/zend_alloc.c:
```diff
@@ -415,50 +415,77 @@ stderr_last_error(char *msg)
 
-#ifndef HAVE_MREMAP
-static void *zend_mm_mmap_fixed(void *addr, size_t size)
+static void zend_mm_munmap(void *addr, size_t size)
```
Why were the locations of these method bodies swapped (e.g. munmap used to be below, and zend_mm_map_fixed was at this location on line 419)

This will make resolving merge conflicts when adding in patches harder, and make seeing the git history (e.g. `git blame`) harder.

If you need to call a function that's declared later, you can declare it above the definition (in the same ifdef checks, if any, for _WIN32 functions)

`static void *zend_mm_mmap_fixed(void *addr, size_t size);`

mvorisek replied:
`zend_mm_munmap` should be used in `zend_mm_mmap_fixed` (previously `munmap` and related error handling there was duplicated)

TysonAndre replied:
Yes, I mean

```c
static void zend_mm_munmap(void *addr, size_t size); // declaration of method, e.g. in the top statements of the file

static void *zend_mm_mmap_fixed(void *addr, size_t size) { // definition of method in original location in the file, preserving git history
    /* ... call zend_mm_munmap */
}

static void zend_mm_munmap(void *addr, size_t size) { // definition of method in original location in the file, preserving git history
}
```

mvorisek replied:
the low-levels methods are modified a lot, basically rewritten, I prefer to not declare/prototype static method first, also, no allocation is expected to be done in free func later

TysonAndre commented on Zend/zend_alloc.c:
```diff
@@ -415,50 +415,77 @@ stderr_last_error(char *msg)
+	ZEND_ASSERT(ptr == addr);
 	return ptr;
+}
 #else
```
```suggestion
#else /* Common operating systems (Not 32-bit windows) */
```

mvorisek replied:
non x86/x64

TysonAndre replied:
```suggestion
#else /* Not _WIN32 */
```

https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170 says x86, x64, and ARM, so I assume that's all the platforms windows supports?

mvorisek replied:
yes, no any Windows, added a comment to this long #if/#else

TysonAndre commented on Zend/zend_alloc.c:
```diff
@@ -663,35 +696,47 @@ static zend_always_inline int zend_mm_bitset_is_free_range(zend_mm_bitset *bitse
+#ifdef _WIN32
+	void *ptr = NULL;
+	size_t offset;
+#ifndef _WIN64
```
I'm confused by this. I think the `_WIN32` and `_WIN64` macros are mutually exclusive, and you don't need to add code for `_WIN64`. (WIN32 without an underscore is the macro for the existence of windows 32 apis)

- _WIN64 has support for mmap (not targeting legacy platforms?), and we use mmap/munmap on 64-bit windows

https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170

> `_WIN32` Defined as 1 when the compilation target is 32-bit ARM, 64-bit ARM, x86, or x64. Otherwise, undefined.
>
> `_WIN64` Defined as 1 when the compilation target is 64-bit ARM or x64. Otherwise, undefined.

mvorisek replied:
I used `_WIN32` as before for detecting any Windows (x86/x64), anything to adjust?

TysonAndre replied:
> `_WIN32` Defined as 1 when the compilation target is 32-bit ARM, 64-bit ARM, x86, or x64. Otherwise, undefined.

Oh. I misread that and missed the part about 64-bit and x64.

TysonAndre commented on Zend/zend_alloc.c:
```diff
@@ -663,35 +696,47 @@ static zend_always_inline int zend_mm_bitset_is_free_range(zend_mm_bitset *bitse
+	}
+	zend_mm_munmap(ptr, size + alignment);
+	ptr = zend_mm_mmap_reserve_try((void*)((char*)ptr + offset), size);
+#endif
```
Remove redundant ifndef _WIN64 check?

mvorisek replied:
`_WIN64` is detecting x64 only (`_WIN86` is however for both, probably by some BC reasons Microsoft decised to do so...)

TysonAndre commented on Zend/zend_alloc.c:
```diff
@@ -415,50 +415,77 @@ stderr_last_error(char *msg)
+	return ptr;
+}
+
+#ifndef _WIN64
```
Remove redundant ifndef? Maybe add function name prefixes to make it more obvious this code path only affects win32 (useful when trying to understand the allocator on commonly used platforms)?

Same for zend_mm_mmap_commit (zend_mm_mmap_win32_commit)

mvorisek replied:
there is a comment right below

TysonAndre reviewed (commented):
Oh, right, I forgot about seeing VirtualQuery - the usage seems right at a glance

TysonAndre commented on Zend/zend_alloc.c:
```diff
@@ -415,50 +415,77 @@ stderr_last_error(char *msg)
+		if (GetLastError() != ERROR_INVALID_ADDRESS) {
 #if ZEND_MM_ERROR
-			fprintf(stderr, "\nmunmap() failed: [%d] %s\n", errno, strerror(errno));
+		stderr_last_error("VirtualAlloc() fixed reserve failed");
```
```suggestion
			stderr_last_error("VirtualAlloc() fixed reserve failed");
```

mvorisek replied:
wdym?

TysonAndre replied:
The statements within an `if` statement should have one more tab of indentation than the line with the `if`

mvorisek replied:
👍

TysonAndre commented on Zend/zend_alloc.c:
```diff
@@ -663,35 +696,47 @@ static zend_always_inline int zend_mm_bitset_is_free_range(zend_mm_bitset *bitse
+#ifndef _WIN64
+	/* 32-bit build has limited memory address space and partical deallocation is not supported,
+	 * try to reserve exact aligned size first, GH-9650 */
+	ptr = zend_mm_mmap_reserve(size + alignment);
```
This is new code. I think coding style guidelines allowed for declaring variables after other executed statements for php 8.0 and even older (no need to support old C standards)

`void *ptr = zend_mm_mmap_reserve(size + alignment);`

mvorisek replied:
`#ifndef _WIN64` there is to detect non-x64, ie. x86 only, thus declaration above is needed

TysonAndre commented on Zend/zend_alloc.c:
```diff
@@ -663,35 +696,47 @@ static zend_always_inline int zend_mm_bitset_is_free_range(zend_mm_bitset *bitse
+	size_t offset;
+#ifndef _WIN64
+	/* 32-bit build has limited memory address space and partical deallocation is not supported,
+	 * try to reserve exact aligned size first, GH-9650 */
```
The original code (and mmap code path) had `void *ptr = zend_mm_mmap(size);` and would return early if aligned and free it if that failed. I think it should continue doing that, and only do this as a fallback - was this still planned?

I assume it's faster to make 1 call to the allocator in the fast path than 3 (alloc + free + alloc at specific address)

Additionally, most of the time I'd expect allocated blocks to be allocated aligned to the 2MB boundary for long-running programs?

- e.g. if it's allocating blocks from the bottom up, then if the previous block was aligned to 2MB and of size 2MB, then the next free block of size 2MB will also usually be aligned

mvorisek replied:
intended and tested - `zend_mm_mmap(size)` returns almost always an unaligned virtual address, thus in the end, the original approach was doing much more allocation calls

for x86 we do it to address https://github.com/php/php-src/pull/9721#issuecomment-1277843700

TysonAndre replied:
> zend_mm_mmap(size) returns almost always an unaligned virtual address (on windows), thus in the end, the original approach was doing much more allocation calls

Could you document that it's almost always unaligned on Windows?

Does that just apply to startup, or do you also mean that happens in subsequent allocations after the first in long-running php processes (In the original `PHP-8.0` branch, not this branch)

mvorisek replied:
`zend_mm_mmap(size)` is never tried to be allocated first for these reasons

cmb69 reviewed (commented):
I don't think it is ever a good idea to do such considerable changes to a *stable* version, especially when it is about an occassionally occurring bug on an almost irrelevant architecture (32bit Windows).

mvorisek commented:
this PR fixes memory allocation in reliable way, Windows x64 is affected as well only it takes much more iterations to reproduce

also the "alloc more -> free -> hope still free -> alloc exact fixed" idea is wrong for ZTS Windows, as same process Windows threads share virtual address space

given these facts, it is a clear bugfix which must go into 8.0

cmb69 commented:
> also the "alloc more -> free -> hope still free -> alloc exact fixed" idea is wrong for ZTS Windows, as same process Windows threads share virtual address space

That is not Windows specific, though (albeit ZTS is probably rarely used on non Windows systems). However, while the patch reduces the amount of race conditions, it does not completely prevent them. To make sure that there are never race conditions, we'd need to employ a mutex, but that *might* not be feasible for performance reasons.

 > given these facts, it is a clear bugfix which must go into 8.0

No. There are always trade-offs (otherwise there would not be such a thing like "WONTFIX"). And for any bugfix targeting a stable version we *need* to be reasonably sure that it doesn't cause a regression. I'm not sure in this case (patch is almost impossible for me to parse), although it seems to me that the splitting reservation/commit makes sense.

Anyhow, I think the most relevant issue we have here is the waste of reserved memory; [like @TysonAndre already said](https://github.com/php/php-src/pull/9721#issuecomment-1274677211) this could be used by employing `VirtualAlloc2()`, and in my opinion we should [use that ("master" only) and load it dynamically](https://github.com/microsoft/mimalloc/blob/master/src/os.c) to avoid requiring Windows 10, but to still be able to use in most cases.

mvorisek commented:
> However, while the patch reduces the amount of race conditions, it does not completely prevent them.

this patch changes Windows allocation only and for Windows it fixes the problem absolutely without any race conditions, even no race conditons for ZTS anymore

the used memory allocator is the same, only we split the reserve and commit steps into a separate calls (previously there were even more alloc calls)

the changes might look complex at the first sight, but it is because linux/mmap does not support memory reservation without commit, thus the static methods were required to be split, renamed and adjusted

cmb69 commented on Zend/zend_alloc.c:
```diff
@@ -471,33 +498,39 @@ static void *zend_mm_mmap(size_t size)
-	if (munmap(addr, size) != 0) {
-#if ZEND_MM_ERROR
-		fprintf(stderr, "\nmunmap() failed: [%d] %s\n", errno, strerror(errno));
+	/* MAP_FIXED leads to discarding of the old mapping, so it can't be used. */
+	void *ptr = mmap(addr, size, PROT_READ | PROT_WRITE, flags /*| MAP_POPULATE | MAP_HUGETLB*/, ZEND_MM_FD, 0);
```
This is not a Windows change.

cmb69 commented on Zend/zend_alloc.c:
```diff
@@ -663,35 +696,47 @@ static zend_always_inline int zend_mm_bitset_is_free_range(zend_mm_bitset *bitse
+	if (offset != 0) {
+		offset = alignment - offset;
+	}
+	zend_mm_munmap(ptr, size + alignment);
+	ptr = zend_mm_mmap_win_reserve_fixed((void*)((char*)ptr + offset), size);
```
Here is still a race condition.

mvorisek replied:
to address https://github.com/php/php-src/pull/9721#issuecomment-1277843700 limited memory space on Windows x86, original approach is used, if it fails, the new non-race code is used for the allocation

cmb69 commented:
If we merged that into PHP-8.0, it would be released with PHP 8.0.26 (scheduled for 2022-11-24), so likely the last regular revision. If it would cause issues for anyone, they likely would not be able to update to any of the following security releases. That seems to be *very* bad.

mvorisek commented:
here is a repo https://github.com/mvorisek/php-src/commit/fbf4cfd221f06bed8e2d5f84174f36992f7e2a31 for debug (with nice alloc logging) and results with the original appoach, when zend_alloc from this PR is used, all MM problems are gone

> If we merged that into PHP-8.0, it would be released with PHP 8.0.26 (scheduled for 2022-11-24), so likely the last regular revision. If it would cause issues for anyone, they likely would not be able to update to any of the following security releases. That seems to be very bad.

Thankfully, we can fix it before the support ends :) and in case of some very bad findings it can be reverted anytime

derickr commented:
I am with @cmb69 in saying that this can't go into PHP 8.0.

There is only one more release (8.0.26, to be released 2022-11-24). If this fix ends up causing problems in that last release, then there is no room for fixing that ... as the support period has already ended. This risk is too high.

This should go into 8.1 or later only, if approved.

mvorisek commented:
Retargetted to PHP 8.1. Thankfully, there was no merge conflict.

TysonAndre commented:
The rebase seems like it was performed correctly - I'm fine with changing the fatal error format and with targeting php 8.1, and my previous code review comments were addressed

dstogov reviewed (changes requested):
The patch is a bit messed and hard to read. It reorders functions and spreads the logic over them. I suppose, you liked to do something like the following.

```diff
diff --git a/Zend/zend_alloc.c b/Zend/zend_alloc.c
index dfdc4e2bb4..42fa095f66 100644
--- a/Zend/zend_alloc.c
+++ b/Zend/zend_alloc.c
@@ -680,18 +680,42 @@ static void *zend_mm_chunk_alloc_int(size_t size, size_t alignment)
 
 		/* chunk has to be aligned */
 		zend_mm_munmap(ptr, size);
-		ptr = zend_mm_mmap(size + alignment - REAL_PAGE_SIZE);
 #ifdef _WIN32
+		ptr = VirtualAlloc(NULL, size + alignment, MEM_RESERVE, PAGE_READWRITE);
+		if (ptr == NULL) {
+# if ZEND_MM_ERROR
+			stderr_last_error("VirtualAlloc() reserve failed");
+# endif
+			return NULL;
+		}
 		offset = ZEND_MM_ALIGNED_OFFSET(ptr, alignment);
-		zend_mm_munmap(ptr, size + alignment - REAL_PAGE_SIZE);
-		ptr = zend_mm_mmap_fixed((void*)((char*)ptr + (alignment - offset)), size);
+		if (VirtualFree(ptr, 0, MEM_RELEASE) == 0) {
+# if ZEND_MM_ERROR
+			stderr_last_error("VirtualFree() failed");
+# endif
+		}
+		ptr = VirtualAlloc((void*)((char*)ptr + (alignment - offset)), size, MEM_RESERVE, PAGE_READWRITE);
+		if (ptr == NULL) {
+# if ZEND_MM_ERROR
+			stderr_last_error("VirtualAlloc() fixed reserve failed");
+# endif
+			return NULL;
+		}
 		offset = ZEND_MM_ALIGNED_OFFSET(ptr, alignment);
 		if (offset != 0) {
-			zend_mm_munmap(ptr, size);
+			VirtualFree(ptr, 0, MEM_RELEASE);
+			return NULL;
+		}
+		ptr = VirtualAlloc(ptr, size, MEM_COMMIT, PAGE_READWRITE);;
+		if (ptr == NULL) {
+# if ZEND_MM_ERROR
+			stderr_last_error("VirtualAlloc() commit failed");
+# endif
 			return NULL;
 		}
 		return ptr;
 #else
+		ptr = zend_mm_mmap(size + alignment - REAL_PAGE_SIZE);
 		offset = ZEND_MM_ALIGNED_OFFSET(ptr, alignment);
 		if (offset != 0) {
 			offset = alignment - offset;
```

I didn't check this. Could you please rewrite the patch in a similar minimal way.

TysonAndre commented:
> My personal preference is to keep the git blame history clean and keep methods in their original locations, and to use forward function declarations `static typename xyz(...);` when methods need to call below methods. I'd hope this would make merge conflicts less likely. not a blocker if other reviewers don't object

Other reviewers have objected.

mvorisek commented:
I have split the commit into two - one for easy to check/follow refactoring and then commit with the actual functional changes.

dstogov commented:
> I have split the commit into two - one for easy to check/follow refactoring and then commit with the actual functional changes.

If it's possible to fix a problem with 1 screen diff. It doesn't make a lot of sense to do the same with 10 times bigger patch. (introducing functions used once, reordering them and so on). The less - the better, especial if we target this into an old stable version and don't like to introduce new problems.

dstogov commented:
@mvorisek in case your patch would clean, I would approve it on last week. Today I spent an hour trying to read it, understand all the changes and don't miss anything. I failed. I understand, it's clear for you, because you already worked on it for a while, but for others it would be easier to rewrite this (without noise). 

- logic improvement and re-factoring are good for development branch
- it's important to keep the readable history of changes for core sub systems  
- agree, it's better to move the first allocation attempt in zend_mm_chunk_alloc_int() under #ifndef _WIN32 
- I don't know why static methods are better for Windows (if they perform syscalls anyway)
- I think, it's better to handle WIN32 and WIN64 in the same way (I'm not completely sure why WIN64 should be handled differently)
- we don't need VirtualQuery. I suppose it comes from an attempt of reusing old code in zend_mm_chunk_alloc_int() or over-design
- I don't care about single-line changes in error messages

I may still misunderstand some details of your patch

mvorisek commented:
>     * logic improvement and re-factoring are good for development branch

I did the refactoring as the new fixed code can be then reviewed much more easier.

The 1st commit can be reviewed by moving the methods locally in text editor and then comparing the result. There are very few changes actually.

>     * I don't know why static methods are better for Windows (if they perform syscalls anyway)

There are static methods: `zend_mm_mmap_win_reserve`, `zend_mm_mmap_win_reserve_fixed`, `zend_mm_mmap_win_commit` for consitency with static methods like `zend_mm_mmap` for linux.

`zend_mm_mmap_win_reserve` is even used more than once.

>     * I think, it's better to handle WIN32 and WIN64 in the same way (I'm not completely sure why WIN64 should be handled differently)

It is handled the same, only on x86 we try allocation subject to race condition first. If it fails, we continue with unified x64/x86 allocation.

>     * we don't need VirtualQuery. I suppose it comes from an attempt of reusing old code in zend_mm_chunk_alloc_int() or over-design

It is needed, as the commited memory start can be different vs. the reserved memory start. To release all/reserved memory, we need to query `VirtualQuery`.

Please review the commits separately, the 1st one is a preparation for the 2nd one, 1st one can be reviewed easily, 2nd one should be obvious quickly.

dstogov commented on Zend/zend_alloc.c:
```diff
@@ -660,28 +697,47 @@ static zend_always_inline int zend_mm_bitset_is_free_range(zend_mm_bitset *bitse
+	ptr = zend_mm_mmap_win_reserve_fixed((void*)((char*)ptr + offset), size);
+#endif
+	if (ptr == NULL) {
+		ptr = zend_mm_mmap_win_reserve(size + alignment);
```
On Win64 you reserve "size+alignment" then commit "size". Is this intended?
I suppose this way, you reserve two times more memory when it's necessary.
See for example how jemalloc does the alignment https://github.com/jemalloc/jemalloc/blob/dev/src/pages.c#L203

mvorisek replied:
the lib/link you provided has the same issue, memory allocation can fail (after unmap and then map with fixed addr) and the app has to deal with it

with 32-build I can reproduce this very consistently in about 1 in 10k tests

dstogov replied:
I hardly believe this is the right approach and jemalloc and others are wrong.

@cmb69 @weltling can you find some Microsoft expert, that can confirm that the following code is the right solution to allocate aligned memory blocks. (e.g. 2MB aligned block of size 2MB).
```c
void *ptr = VirtualAlloc(NULL, size + alignment, MEM_RESERVE, PAGE_READWRITE);
size_t offset = ((size_t)ptr) & (alignment - 1);
ptr = VirtualAlloc((void*)((char*)ptr + (alignment - offset)), size, MEM_COMMIT, PAGE_READWRITE);
```

cmb69 replied:
I don't think that I can find any expert, but from what I can tell, the proper solution to allocate an aligned memory block would be using [VirtualAlloc2](https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc2) (where the desired alignment can be specified). However, that API is only available as of Windows Server 2016/Windows 10, so can't be used by us (or at least we need a fallback).

So, in lack of that API, the code above seems to be the next best solution. It is actually similar to what we are doing now, although we currently don't split the reservation and the commit. Note that the `VirtualAlloc(NULL, …)` call is unlikely to give a 2MiB aligned memory address; the only guarantee is that it is aligned on the page size (which is 64KiB for me). So we do the re-allocation most of the time anyway to force the 2MiB alignment. There is likely a small performance gain if we don't commit the `NULL` allocation, because that zeroes the memory, while only reserving it won't touch the memory contents. The obvious drawback of splitting reservation/commit would be that we need to track the reserved address (or fetch it via `VirtualQuery()`).

I wonder, though, why we even try to enforce any alignment larger than the page size. From what I can tell, jemalloc would also do that, but it seems that they use a smaller page size than 2MiB (unless explicitly configured).

All in all, I see potential for improvement regarding the allocation, but it would make sense to actually do some performance measurements, but these experiments might better target the "master" branch only.

mvorisek replied:
I did more research and official Microsoft mimalloc is using exactly the same approach.

aligned alloc reserve: https://github.com/microsoft/mimalloc/blob/v2.0.6/src/os.c#L766 and commit: https://github.com/microsoft/mimalloc/blob/v2.0.6/src/os.c#L776

free https://github.com/microsoft/mimalloc/blob/v2.0.6/src/os.c#L366

dstogov replied:
OK. Now I see that you are not alone who use this trick. (I'm not sure if it good or not. jemalloc is more known/used library and it uses simple loop in case of failure. Personally, I would prefer jemalloc version).

BTW: You do the trick only for Win32 (not Win64), but call VirtulQuery for both Win32 and Win64. Right? It's useless on Win64. Right?

mvorisek replied:
reserve more + commit only aligned size is used on both x86/x64

on x86, the original approach (subject to race condition) is tried once first to try to reserve the address space more densely, on x64 not needed and will imply only more syscalls

please note `#ifdef _WIN32` matches x86/x64 and `#ifdef _WIN32` x64 only

cmb69 replied:
> please note #ifdef _WIN32 matches x86/x64 and #ifdef _WIN64 x64 only

Right, but also for ARM.

TysonAndre replied:
> I wonder, though, why we even try to enforce any alignment larger than the page size. From what I can tell, jemalloc would also do that, but it seems that they use a smaller page size than 2MiB (unless explicitly configured).

I was looking into the allocator last year or so - it's because of https://github.com/php/php-src/blob/98bdb7f99b60d4976797e9764c1c265413ab7915/Zend/zend_alloc.c#L40-L45

Given a pointer returned by emalloc, you can figure out which 2MiB chunk (or the address of the chunk header) as part of by a single bitwise operation 

https://github.com/php/php-src/blob/98bdb7f99b60d4976797e9764c1c265413ab7915/Zend/zend_alloc.c#L203-L225

dstogov replied:
Right. Sorry, I was confused by ``#ifndef _WIN64`` block in zend_mm_chunk_alloc_int().
Why it's necessary only for 32-bit Windows? (actually, it implements the jemalloc approach)

TysonAndre replied:
mvorisek mentioned why he'd went with this here https://github.com/php/php-src/pull/9721#issuecomment-1277890477

In practice, it sounds like 64-bit Windows doesn't have this issue, and the VirtualFree+VirtualAlloc within the freed region works consistently on 64-bit windows but not 32-bit windows

(I don't have windows installed to test)

> > The approach of a finite number of retries I'd mentioned earlier seems like the best option.
>
> several tries of "alloc + dealloc + smaller aligned alloc" fails once the 1st failed, that is almost expected, we will have to store the allocated parts first... and still only a probabilistic "almost never to fail" solution

dstogov replied:
I don't understand. 
- Win32 first tries jemalloc approcah resevrce(sizealigned)+free+aligned_reserve(size)+commit,
then fallback to reserve(size+aligned)+aligned_commit(size),
- Win64 only the second fallback reserve(size+aligned)+aligned_commit(size)
Why Win64 don't use the jemalloc approch (it should work). Or why Win32 use it, if it doesn't work?

dstogov replied:
I think this ``#ifndef _WIN64`` should be removed. Only #ifndef itself (not the code below).

mvorisek replied:
in 32-bit Windows the virtual memory space is limited to 2GB, so we try the racy allocation first which might fail and fallback to the fixed/x64 allocation, but if it succeed, only the exact memory range is reserved - in 64-bit, we do not need it, we can simply reserve more (and still commit only the aligned exact size)

this PR is not subject to any race condition (for both x86/x64), fixes all the memory random failures and I did not notice any performance nor other problems

also, similar solution is used by very reputable mimalloc memory allocator which is developer directly by Microsoft and has very good performance characteristics

mvorisek replied:
> I think this `#ifndef _WIN64` should be removed. Only #ifndef itself (not the code below).

`#ifndef _WIN64` means not 64-bit Windows, ie. 32-bit Windows only (if wrapped in `#ifdef _WIN32`), I double checked the code, the `#ifndef _WIN64` seems correct

dstogov replied:
yes, Why this code should work only for 32-bit Windows? This means we will waste address space on Win64 only. 

Do you have race condition problems on ZTS build only?

dstogov replied:
Look, you have a problem on 32-bit Windows, and you fix this adding fallback code for 32-bit Windows, but you also completely change the behavior for 64-bit Windows to use new fallback instead of old approach.

mvorisek replied:
we will waste a few atoms in universe (64-bit virtual address space) :D - is that a problem? bigger reservation should be fine, we only reserve larger range, but the actual commited physical memory size is always the same/exact

as long as there is no measureable problem (and I do not know any), I prefer to not even try the racy approach first for x64, as it needs more syscalls + error handling using global `GetLastError()`

ZTS - no, I can reproduce the problem consistently also with NTS

dstogov replied:
I can't understand this. This "racy approach" worked for ages, and you tell it fails with 0.1% probability on Win32 only (I've never seen this kind of failures). Why do you change Win64 related behavior, and do it in a way that is really hard to catch. This patch is not just a fix!

Can you rewrite the fix that only adds the new fallback (I don't care for win32 only or for both). And do it in a compact readable way?

mvorisek replied:
problem is present on x86 and x64, on x64 it is much more rare as the addr collisions are much more rare

the original "racy" approach is not needed at all, it is there only to try to reserve exact memory size first to not reduce limited 2GB address space

so simply the original "racy" approach is problematic more or less on all Windows

> can you rewrite the fix that only adds the new fallback (I don't care for win32 only or for both). And do it in a compact readable way?

the approach in this PR uses minimal number of syscalls, it cannot be simplified

I would say this path is a fix, as the original approach is wrong/unfixable, no public API is changed and the allocator still uses the same `VirtualAlloc` function.

`#ifndef _WIN64` can be removed in theory, but I prefer to not do so, as it would only increase the number of syscalls, reserving "size+aligned" in 64-bit address space should not be worse than "size" only - is this the problem you have with this PR? If yes, I propose to remove the `#ifndef _WIN64` and add them for `master`.

dstogov replied:
I gave up. I can't waste more time. I don't approve this, but fell free to commit if others approve. I won't object. (May be I'm a stubborn donkey).

TysonAndre replied:
>I can't understand this. This "racy approach" worked for ages, and you tell it fails with 0.1% probability on Win32 only (I've never seen this kind of failures).

It's an issue I've been seeing in a small fraction of ci builds, when starting around a thousand php processes total per 32-bit execution (in a plugin starting a lot of php processes for `--syntax-check`, where each build included 2 jobs using 32-bit php for the test suites of the linked project)

https://ci.appveyor.com/project/TysonAndre/phan/builds/44983646/job/1f74k3nsurgxpdgb (more build failures in the job history, all jobs were running 32 bit on windows because I hadn't gotten around to setting up 32 bits on Linux before that

```
Analyzing files...
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
Can't initialize heap: [0x000001e7] Attempt to access invalid address
░░░░░  54 / 619 (  8%) 409MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 108 / 619 ( 17%) 426MB
```

After seeing the cause of the failure I migrated AppVeyor to use 64 bit php and added 32-bit linux tests to work around this https://github.com/phan/phan/commit/35f0e26e49ced83dc1fd8e9fbf771f60dd927e3a

TysonAndre replied:
> Can you rewrite the fix that only adds the new fallback (I don't care for win32 only or for both). And do it in a compact readable way?

I agree with that review comment.

I would have gone with it if Dmitry Stogov could tell if it was correct and agreed with the approach, but I'm not confident in this (e.g. CI tests are different from a long running web server).

- My earlier suggestion of keeping the method order unchanged and adding C function declarations would make this smaller and easier to review

It seems safer and easier to maintain to change only on affected platforms  to avoid unexpectedly changing behavior of this (e.g. to guarantee it only affects Windows), especially in 8.1, which already has stable releases



The null checks seem beneficial to ensure the error codes are from the first syscall that failed, but smaller PRs would be easier to review or bisect if anything did go wrong

mvorisek replied:
@TysonAndre thank you for rephrasing what I need to do ❤️

I will update this PR to be a pure fallback with minimal LOC changes needed.

For master, I will then submit another PR with the current impl. which is not a "fallback" (for non-x86) but is optimized to require less syscalls.

mvorisek replied:
done

dstogov commented on Zend/zend_alloc.c:
```diff
@@ -419,6 +419,14 @@ stderr_last_error(char *msg)
 static void zend_mm_munmap(void *addr, size_t size)
 {
 #ifdef _WIN32
+	MEMORY_BASIC_INFORMATION mbi;
+	if (VirtualQuery(addr, &mbi, sizeof(mbi)) == 0) {
+#if ZEND_MM_ERROR
+		stderr_last_error("VirtualQuery() failed");
+#endif
+	}
+	addr = mbi.AllocationBase;
+	
```
I understand that this VirtualQuery() is need to handle unaligned reserved blocks, but it's better to reserve aligned block in first place. I don't know where you find this trick. Is it recommended by MSDN?

mvorisek replied:
I came to the solution after huge research, all Windows memory allocators are wrappers around VirtualAlloc/VirtualFree/VirtualQuery API

mvorisek commented:
I have reworked this PR to be a minimalistic bugfix and a fallback solution only when https://github.com/php/php-src/issues/9650 is hit.

When https://github.com/php/php-src/issues/9650 is hit, the allocated memory might be up to the alignment size larger, but such problem is hit in about 0.01% alloc calls on x86 Windows only, so it should be fine.

For easier review, please follow the individual commits, if possible, please keep the commits unsquashed when merging.

dstogov commented on Zend/zend_alloc.c:
```diff
@@ -682,14 +699,28 @@ static void *zend_mm_chunk_alloc_int(size_t size, size_t alignment)
+				ptr = (void*)((char*)ptr + alignment - offset);
+			}
+        } else {
+			offset = ZEND_MM_ALIGNED_OFFSET(ptr, alignment);
+			if (offset != 0) {
+				zend_mm_munmap(ptr, size);
+				return NULL;
+			}
```
I'm not sure if this check is necessary, because you already request the fixed aligned segment. Anyway, this is not a big problem.

mvorisek replied:
removed - good catch, this was there originally, but it will never be executed, also there is `ZEND_ASSERT(ptr == addr)` newly in `zend_mm_mmap_fixed`

dstogov commented on Zend/zend_alloc.c:
```diff
@@ -416,11 +416,49 @@ stderr_last_error(char *msg)
+{
+#ifdef _WIN32
+	MEMORY_BASIC_INFORMATION mbi;
+	if (VirtualQuery(addr, &mbi, sizeof(mbi)) == 0) {
```
It may make sense to call VirtualFree() first and than fall-back to VirtualQuery()+VirtualFree() like mi_os_mem_free().
See https://github.com/microsoft/mimalloc/blob/v2.0.6/src/os.c#L359..L372

mvorisek replied:
I am aware they do it this way, but I query `VirtualQuery` always to prevent the modification of global error code (`err`).

dstogov replied:
Now the allocation of "unaligned" block occurs as a fall-back for case when aligned block can't be allocated. I suspect we will have more aligned blocks then unaligned and therefore single VirualFree() for the most often case might be better.

mvorisek replied:
done

dstogov replied:
@cmb69 could you please take a final look and merge this

dstogov reviewed (approved):
The patch might be smaller, but now it's readable and looks right.
(I didn't test, just looked over the code).

cmb69 commented:
Thank you all!
